### PR TITLE
Add an App struct and use gtk::Application::startup only for creating…

### PR DESCRIPTION
**Pull request by sdroege**
… the menu and window

The window is shown/presented only in gtk::Application::activate, which
also has the side-effect that when a second instance of the application
is started only the window of the first instance is brought to the
foreground.

----

CC @GuillaumeGomez

### CODE REVIEW

File: src/main.rs [hunk 1]
```diff
@@ -29,9 +56,9 @@ fn build_menu(application: &gtk::Application, window: &gtk::ApplicationWindow) {
 
     // Connect the about menu entry click event.
     let about = gio::SimpleAction::new("about", None);
-    let weak_window = window.downgrade();
+    let weak_application = application.downgrade();
```
Comment (GuillaumeGomez) [inline src/main.rs:59]: I just thought that we could greatly improve the `upgrade_weak` macro to make it look like the `clone` macro (taking arguments and calling `upgrade` on all given arguments).

File: src/main.rs [hunk 2]
```diff
@@ -40,16 +67,19 @@ fn build_menu(application: &gtk::Application, window: &gtk::ApplicationWindow) {
...
     });
+
     let quit = gio::SimpleAction::new("quit", None);
-    let weak_window = window.downgrade();
+    let weak_application = application.downgrade();
     quit.connect_activate(move |_, _| {
-        let window = upgrade_weak!(weak_window);
-        window.destroy();
+        let application = upgrade_weak!(weak_application);
+        application.quit();
```
Comment (GuillaumeGomez) [inline src/main.rs:82]: Oh better indeed! Never paid attention to this...

File: src/main.rs [hunk 3]
```diff
@@ -93,19 +124,42 @@ fn build_ui(application: &gtk::Application) {
...
+    application.connect_activate(move |_| {
+        let app = upgrade_weak!(app_weak);
+        let inner = app.0.borrow();
+        // We only show our window here once the application
+        // is activated. This means that when a second instance
+        // is started, the window of the first instance will be
+        // brought to the foreground
+        if let Some(ref main_window) = inner.main_window {
+            main_window.show_all();
+            main_window.present();
```
Comment (GuillaumeGomez) [inline src/main.rs:155]: Makes more sense actually.

File: src/main.rs (hunk 1 above)
Comment (sdroege) [reply src/main.rs:59]: I have that on my TODO list, a combined macro that does clone or downgrade based on a marker in the signature and also allows renaming things (so you can pass `self` from outside the closure into the closure, where `self` would be illegal).

Comment (GuillaumeGomez) [reply src/main.rs:59]: That'd be amazing! We definitely need a macro crate in gtk-rs for such things.

**Pull request merged**